File: js/test/test.BBBrowserExtensionInjector.ts
import assert from "node:assert/strict";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { describe, it } from "vitest";

import { BBBrowserExtensionInjector } from "../src/injector/BBBrowserExtensionInjector.js";
import { ModCDPClient } from "../src/client/ModCDPClient.js";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const EXTENSION_PATH = path.resolve(HERE, "..", "..", "dist", "extension");
const hasBrowserbaseEnv = Boolean(process.env.BROWSERBASE_API_KEY?.trim());

describe("BBBrowserExtensionInjector", () => {
  it("uses configured extension id", async () => {
    const injector = new BBBrowserExtensionInjector({ injector_extension_id: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" });

    await injector.prepare();

    assert.equal(injector.getLauncherConfig().injector_extension_id, "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa");
  });

  it("does not upload when no extension path or id is configured", async () => {
    const injector = new BBBrowserExtensionInjector();

    await injector.prepare();

    assert.deepEqual(injector.getLauncherConfig(), {});
    assert.equal((injector as unknown as { zip_path: string | null }).zip_path, null);
  });
});

describe.skipIf(hasBrowserbaseEnv)("BBBrowserExtensionInjector without credentials", () => {
  it("requires BROWSERBASE_API_KEY when extension upload is needed", async () => {
    const injector = new BBBrowserExtensionInjector({ injector_extension_path: EXTENSION_PATH });

    try {
      await assert.rejects(() => injector.prepare(), /BROWSERBASE_API_KEY/);
      assert.equal((injector as unknown as { cleanup: unknown }).cleanup, null);
    } finally {
      await injector.close();
    }
  });
});

describe.skipIf(!hasBrowserbaseEnv)("BBBrowserExtensionInjector", () => {
  it(
    "uploads the real extension and launches a Browserbase browser with it installed",
    { timeout: 180_000 },
    async () => {
      const cdp = new ModCDPClient({ launcher: {
          launcher_mode: "bb",
          launcher_options: {
            browserbase_project_id: process.env.BROWSERBASE_PROJECT_ID,
            timeout: 120,
            ...(process.env.BROWSERBASE_REGION ? { region: process.env.BROWSERBASE_REGION } : {}),
          },
        },
        upstream: { upstream_mode: "ws" }, injector: {
          injector_mode: "inject",
          injector_extension_path: EXTENSION_PATH,
          injector_service_worker_url_suffixes: ["/modcdp/service_worker.js"],
          injector_trust_service_worker_target: true,
        },
      });

      try {
        await cdp.connect();
        assert.equal(cdp.connect_timing?.injector_source, "bb");
        assert.equal(typeof cdp.extension_id, "string");
        const service_worker_url = await cdp.Mod.evaluate({
          expression: "chrome.runtime.getURL('modcdp/service_worker.js')",
        });
        assert.match(String(service_worker_url), /^chrome-extension:\/\/[a-z]{32}\/modcdp\/service_worker\.js$/);
      } finally {
        await cdp.close();
      }
    },
  );
});

File: python/tests/test_BBBrowserExtensionInjector.py
from __future__ import annotations

import os
from pathlib import Path
import unittest
from unittest.mock import patch

from modcdp import ModCDPClient
from modcdp.injector.BBBrowserExtensionInjector import BBBrowserExtensionInjector

HERE = Path(__file__).resolve().parent
EXTENSION_PATH = HERE.parents[1] / "dist" / "extension"


class BBBrowserExtensionInjectorTests(unittest.TestCase):
    def test_prepares_default_packaged_extension_zip_when_path_is_omitted(self) -> None:
        injector = BBBrowserExtensionInjector()
        try:
            with patch.object(injector, "_uploadExtension", return_value="aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa") as upload:
                injector.prepare()
            self.assertTrue(str(injector.options.get("injector_extension_path", "")).endswith("extension.zip"))
            self.assertTrue(str(injector.zip_path or "").endswith("extension.zip"))
            upload.assert_called_once_with(injector.zip_path)
            self.assertEqual(injector.getLauncherConfig(), {"injector_extension_id": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"})
        finally:
            injector.close()

    def test_uploads_real_extension_and_launches_browserbase_browser_with_it_installed(self) -> None:
        if not os.environ.get("BROWSERBASE_API_KEY", "").strip():
            self.fail("BROWSERBASE_API_KEY is required for live Browserbase tests")
        cdp = ModCDPClient(
            launcher={
                "launcher_mode": "bb",
                "launcher_options": {
                    "timeout": 120,
                    **({"region": os.environ["BROWSERBASE_REGION"]} if os.environ.get("BROWSERBASE_REGION") else {}),
                },
            },
            upstream={"upstream_mode": "ws"},
            injector={
                "injector_mode": "inject",
                "injector_extension_path": str(EXTENSION_PATH),
                "injector_service_worker_url_suffixes": ["/modcdp/service_worker.js"],
                "injector_trust_service_worker_target": True,
            },
        )

        try:
            cdp.connect()
            self.assertEqual(cdp.connect_timing.get("injector_source") if cdp.connect_timing else None, "bb")
            self.assertIsInstance(cdp.extension_id, str)
            service_worker_url = cdp.Mod.evaluate(expression="chrome.runtime.getURL('modcdp/service_worker.js')")
            self.assertRegex(str(service_worker_url), r"^chrome-extension://[a-z]{32}/modcdp/service_worker\.js$")
        finally:
            cdp.close()


if __name__ == "__main__":
    unittest.main()
